stdlib/public/runtime/StringComparison.cpp

```diff
@@ -0,0 +1,162 @@
+//===--- StringComparison.cpp ---------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2018 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "swift/Runtime/StringComparison.h"
+#include "swift/Runtime/Config.h"
+#include <cassert>
+#include <cstdint>
+
+#ifdef __APPLE__
+
+#include <simd/simd.h>
+
+static int findDiffIdxSmall(uint8_t *left, uint16_t *right, int count);
+
+namespace {
+static constexpr simd_packed_uchar16 zero = {0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+                                             0x0, 0x0, 0x0, 0x0, 0x0, 0x0,
+                                             0x0, 0x0, 0x0, 0x0};
+} // end anonymous namespace
+
+/// This is equivalent to _mm_unpacklo_epi8(x, zero) but written in a manner
+/// that should work across platform.
+static simd_packed_ushort8 unpackLower(simd_packed_uchar16 input) {
+  return (simd_packed_ushort8)__builtin_shufflevector(
+      input, zero, 0, 16, 1, 17, 2, 18, 3, 19, 4, 20, 5, 21, 6, 22, 7, 23);
+}
+
+/// This is equivalent to _mm_unpackhi_epi8(x, zero) but written in a manner
+/// that should work across platform.
+static simd_packed_ushort8 unpackUpper(simd_packed_uchar16 input) {
+  return (simd_packed_ushort8)__builtin_shufflevector(input, zero, 8, 24, 9, 25,
+                                                      10, 26, 11, 27, 12, 28,
+                                                      13, 29, 14, 30, 15, 31);
+}
+
+/// This is equivalent to _mm_unpacklo_epi8(lhs, zero) == rhs
+static simd_packed_short8 performComparisonLow(simd_packed_uchar16 lhs,
+                                               simd_packed_ushort8 rhs) {
+  return unpackLower(lhs) != rhs;
+}
+
+/// This is equivalent to _mm_unpackhi_epi8(lhs, zero) == rhs
+static simd_packed_short8 performComparisonHigh(simd_packed_uchar16 lhs,
+                                                simd_packed_ushort8 rhs) {
+  return unpackUpper(lhs) != rhs;
+}
+
+namespace {
+static constexpr int ELTS_PER_ROUND = 64;
+
+// We only want to vectorize if we have more than a page in
+// size. NUM_ELTS_PER_VECTORIZATION provides that.
+static constexpr int NUM_ELTS_FOR_VECTORIZATION = ELTS_PER_ROUND * 4;
+} // end anonymous namespace
+
+// *TODO* use a template to make this autoupgrade to use AVX when -mavx is
+// *passed in.
```

Comment:
Three things to note here:
- we probably don't want to use AVX[2|512] because the expected trip count for these is going to be fairly short, and we don't want to eat the clock speed penalty for surrounding code.
- If you write this sort of thing in terms of wider-than-base-hardware vectors (e.g. uchar32 or 64), you'll get unrolling when targeting older hardware (no AVX) and wider vectors when targeting AVX without needing two versions.
- 64B is probably too aggressive unrolling given the vaguely power-law distribution of string lengths, and the fact that when mismatches occur, they will occur early. I would probably use 32.